index/segment/fs/encoding/fields/README.md

```diff
@@ -0,0 +1,128 @@
+
+Two files are used to represent the stored fields in a segment. The data file contains
+the stored fields for each document in the segment. The index file contains, for each
+document, its corresponding offset in the data file.
+
+## Data File
+
+The data file contains the stored fields for each document. The documents are stored
+serially.
+
+```
+┌───────────────────────────┐
+│ ┌───────────────────────┐ │
+│ │      Document 1       │ │
+│ ├───────────────────────┤ │
+│ │          ...          │ │
+│ ├───────────────────────┤ │
+│ │      Document n       │ │
+│ └───────────────────────┘ │
+└───────────────────────────┘
+```
+
+### Document
+
+Each document is composed of an ID and its stored fields. The ID is a sequence of valid
+UTF-8 bytes and it is encoded first by encoding the length of the ID, in bytes, as a
+variable-sized unsigned integer and then encoding the actual bytes which comprise the ID.
+Following the ID are the fields. The number of fields in the document is encoded first as
+a variable-sized unsigned integer and then the stored fields themselves are encoded.
+
+```
+┌───────────────────────────┐
+│ ┌───────────────────────┐ │
+│ │     Length of ID      │ │
+│ │       (uvarint)       │ │
+│ ├───────────────────────┤ │
+│ │                       │ │
+│ │          ID           │ │
+│ │        (bytes)        │ │
+│ │                       │ │
+│ ├───────────────────────┤ │
+│ │   Number of Fields    │ │
+│ │       (uvarint)       │ │
+│ ├───────────────────────┤ │
+│ │                       │ │
+│ │        Field 1        │ │
+│ │                       │ │
+│ ├───────────────────────┤ │
+│ │                       │ │
+│ │          ...          │ │
+│ │                       │ │
+│ ├───────────────────────┤ │
+│ │                       │ │
+│ │        Field n        │ │
+│ │                       │ │
+│ └───────────────────────┘ │
+└───────────────────────────┘
+```
+
+#### Stored Field
+
+Each stored field is composed of a name and a value. The name and value are a sequence of
+valid UTF-8 bytes and they are stored by encoding the length of the name (value), in bytes,
+as a variable-sized unsigned integer and then encoding the actual bytes which comprise the
+name (value). The name is encoded first and the value second.
+
+```
+┌───────────────────────────┐
+│ ┌───────────────────────┐ │
+│ │  Length of Field Name │ │
+│ │       (uvarint)       │ │
+│ ├───────────────────────┤ │
+│ │                       │ │
+│ │      Field Name       │ │
+│ │        (bytes)        │ │
+│ │                       │ │
+│ ├───────────────────────┤ │
+│ │ Length of Field Value │ │
+│ │       (uvarint)       │ │
+│ ├───────────────────────┤ │
+│ │                       │ │
+│ │      Field Value      │ │
+│ │        (bytes)        │ │
+│ │                       │ │
+│ └───────────────────────┘ │
+└───────────────────────────┘
+```
+
+## Index File
+
+The index file contains, for each postings ID in the segment, the offset of the corresponding
```

Comment:
could you add the caveat that offsets set to maxuint64 or w/e indicates the doc doesn't exist

Reply:
I'd prefer to leave that out until we actually need it. Seems like an unnecessary complexity at this point and it's easy to add in the future.

Reply:
Talked offline, will add.